src/core/lib/surface/server_call.cc

```diff
@@ -119,93 +119,108 @@ void ServerCall::CommitBatch(const grpc_op* ops, size_t nops, void* notify_tag,
-        grpc_slice_buffer_swap(
-            &op.data.send_message.send_message->data.raw.slice_buffer,
-            send.c_slice_buffer());
-        auto msg = arena()->MakePooled<Message>(std::move(send), op.flags);
-        return [this, msg = std::move(msg)]() mutable {
-          return call_handler_.PushMessage(std::move(msg));
-        };
-      });
-  auto send_trailing_metadata =
-      op_index.OpHandler<GRPC_OP_SEND_STATUS_FROM_SERVER>(
-          [this](const grpc_op& op) {
-            auto metadata = arena()->MakePooled<ServerMetadata>();
-            CToMetadata(op.data.send_status_from_server.trailing_metadata,
-                        op.data.send_status_from_server.trailing_metadata_count,
-                        metadata.get());
-            metadata->Set(GrpcStatusMetadata(),
-                          op.data.send_status_from_server.status);
-            if (auto* details =
-                    op.data.send_status_from_server.status_details) {
-              // TODO(ctiller): this should not be a copy, but we have
-              // callers that allocate and pass in a slice created with
-              // grpc_slice_from_static_string and then delete the string
-              // after passing it in, which shouldn't be a supported API.
-              metadata->Set(GrpcMessageMetadata(),
-                            Slice(grpc_slice_copy(*details)));
-            }
-            CHECK(metadata != nullptr);
-            return [this, metadata = std::move(metadata)]() mutable {
-              CHECK(metadata != nullptr);
-              return [this, metadata = std::move(
-                                metadata)]() mutable -> Poll<Success> {
-                CHECK(metadata != nullptr);
-                call_handler_.PushServerTrailingMetadata(std::move(metadata));
-                return Success{};
-              };
-            };
-          });
-  auto recv_message =
-      op_index.OpHandler<GRPC_OP_RECV_MESSAGE>([this](const grpc_op& op) {
-        return message_receiver_.MakeBatchOp(op, &call_handler_);
-      });
-  auto primary_ops = AllOk<StatusFlag>(
-      TrySeq(AllOk<StatusFlag>(std::move(send_initial_metadata),
-                               std::move(send_message)),
-             std::move(send_trailing_metadata)),
-      std::move(recv_message));
-  if (auto* op = op_index.op(GRPC_OP_RECV_CLOSE_ON_SERVER)) {
-    auto recv_trailing_metadata = OpHandler<GRPC_OP_RECV_CLOSE_ON_SERVER>(
-        [this, cancelled = op->data.recv_close_on_server.cancelled]() {
-          return Map(call_handler_.WasCancelled(),
-                     [cancelled, this](bool result) -> Success {
-                       saw_was_cancelled_.store(true,
-                                                std::memory_order_relaxed);
-                       ResetDeadline();
-                       *cancelled = result ? 1 : 0;
-                       return Success{};
-                     });
+
+  auto commit_with_send_ops = [&](auto send_ops) {
```

Comment:
I'm having trouble following the changes in this file.  I wouldn't have expected any server-side changes here, since this PR is really just implementing the client-side retry functionality.  Can you help me understand what's changing here and why?

Reply:
Oh yeah... so this is doing the trailers-only metadata injection.

In v1 this was a transport concern, and it was super ugly and I'd rather not do that again. Instead, v3 has a notion of "no initial metadata, and therefore no messages and only trailing metadata" and everything (except here) ends up cleaner.

This file is detecting that we have a batch with initial & trailing metadata (and no message), and translating that into the promise operations to send no initial metadata and just the trailing metadata.

Ack on this being ugly and awkward - it's coming from getting the promise types right and relatively efficient mapping from the batches, so when we do a direct call promise api this goes away.

Why now? Because retries bring up the need for robust handling of trailers-only responses.

Reply:
Thanks for the context!  This makes a lot more sense now.